library_generation/utils/generation_config_comparator.py

```diff
@@ -0,0 +1,379 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections import defaultdict
+from enum import Enum
+from typing import Dict
+from typing import List
+
+from library_generation.model.gapic_config import GapicConfig
+from library_generation.model.generation_config import GenerationConfig
+from library_generation.model.library_config import LibraryConfig
+from library_generation.utilities import get_library_name
+
+
+class ChangeType(Enum):
+    GOOGLEAPIS_COMMIT = 1
+    REPO_LEVEL_CHANGE = 2
+    LIBRARIES_ADDITION = 3
+    LIBRARIES_REMOVAL = 4
```

Comment:
Is the scripts going to handle `LIBRARIES_REMOVAL` correctly? IIRC, the scripts is going to skip a library if it is not listed in the `generation_config.yaml`.

Reply:
This PR only creates change types of two configurations, it doesn't specify how to handle each type.

For library removal, I expect `generate_pr_body.py` will create an release note entry like `feat/chore: xxx is removed`. However, the actual code in google-cloud-java needs to be manually removed.

Reply:
Would it be better if we ignore this case and add the `feat/chore: xxx is removed` in the corresponding PR? Or we can think about how to remove libraries automatically, but it would probably be a larger effort.

Reply:
Personally, I think it's better to output this change type here because this is the responsible of the comparator (to produce the change type, but not how to handle each type).

This type can serve as a trigger for automatic removal in further improvement and we don't need to refactor the comparator for this case. WDYT?

Reply:
Sounds reasonable to me. But what is going to be the title for the PR that removes a library? Unless you want to make it a `chore` which would be a little confusing, it would still be duplicated in the release-note. I feel ignoring this case for now is better than documenting this behavior and tell release managers to not use `feat` in the removal PR. We don't have to remove the code, they can be commented out until we figure this out.

Reply:
SG. I'll do some refactor then.

Reply:
I commented out `LIBRARIES_REMOVAL` and `GAPIC_REMOVAL`.

Reply:
I thought `GAPIC_REMOVAL` can be handled automatically?

Reply:
@diegomarquezp after #2572 is merged, can we handle GAPIC removal automatically, i.e., deep-remove regex should work at library level?

Reply:
Deep remove does not remove the entire library but the src folders inside that gapic, grpc and proto "sub libraries". It's meant to replace old code in the sublibraries. It wouldn't automatically delete any sublibrary if it stops being listed in the config yaml.

Reply:
@blakeli0 does this answer your question?